### PR TITLE
feat(sql): add FORCE DROP partition SQL syntax, useful to unblock suspended WAL tables

### LINKED ISSUE #4107: can't drop partition if directory is missing
### To reproduce

1. remove partition directory in server
2. drop partition using webui
3. error throw: "Partition does not exist"

### QuestDB version:

7.3.8b

### OS, in case of Docker specify Docker and the Host OS:

Ubuntu 22.04

### File System, in case of Docker specify Host File System:

ext4

### Full Name:

Tien Nguyen

### Affiliation:

Freelancer

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [X] Yes, I have

### Additional context

It is better to have "force" option for dropping partition already removed (either mistake or data loss) to make db can work again. If facing this situation, db stop working with any queries even outside removed partitions

Thank you

### PR DESCRIPTION
**Pull request by ideoma**
Add support of SQL:

```SQL
ALTER TABLE tbl FORCE DROP PARTITION LIST '2022-02-26T155900-000001'
```

This will remove an exact partition with the directory name `2022-02-26T155900-000001` from the table partition list. This SQL is useful workaround when WAL table is suspended because of missing partition on the disk.

Fixes #4107

### CODE REVIEW

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 72 / 88 (81.82%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/MetadataServiceStub.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|35|47|74.47%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|13|16|81.25%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AbstractOperation.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/OperationDispatcher.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|6|6|100.00%|

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 1]
```diff
@@ -1671,6 +1671,83 @@ public void enableDeduplicationWithUpsertKeys(LongList columnsIndexes) {
         }
     }
 
+    @Override
+    public void forceRemovePartitions(LongList partitions) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/TableWriter.java:1675]: can we have more descriptive parameter name. More appropriate would be `partitionTimestamps`

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 2]
```diff
@@ -1671,6 +1671,83 @@ public void enableDeduplicationWithUpsertKeys(LongList columnsIndexes) {
...
+                if (activePartitionDropped) {
+                    // Need to do this only once
+                    closeActivePartition(false);
+                }
+            }
+
+            firstPartitionDropped |= timestamp == txWriter.getPartitionTimestampByIndex(0);
+            columnVersionWriter.removePartition(timestamp);
+            txWriter.removeAttachedPartitions(timestamp);
+            partitionRemoveCandidates.add(timestamp);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/TableWriter.java:1705]: why are we filling up this list? Contents of the list is not used in this method and there is no comment to explain where is it used and why

File: core/src/main/java/io/questdb/cairo/TableWriter.java (hunk 1 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableWriter.java]: done

File: core/src/main/java/io/questdb/cairo/TableWriter.java (hunk 2 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableWriter.java]: I added a comment:
```Java
            // Add the partition to partition remove list that can be deleted if there are no open readers
            // after the commit
            partitionRemoveCandidates.add(timestamp, txWriter.getPartitionNameTxn(index));
```

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 93 / 96 (96.88%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/MetadataServiceStub.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|51|53|96.23%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AbstractOperation.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/OperationDispatcher.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|6|6|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 112 / 117 (95.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/MetadataServiceStub.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|56|60|93.33%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AbstractOperation.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/OperationDispatcher.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|6|6|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 113 / 118 (95.76%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/MetadataServiceStub.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|56|60|93.33%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AbstractOperation.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/OperationDispatcher.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|6|6|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 124 / 132 (93.94%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/MetadataServiceStub.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV2.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|56|60|93.33%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AbstractOperation.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/OperationDispatcher.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV1.java|1|1|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 134 / 142 (94.37%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/MetadataServiceStub.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV2.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|22|24|91.67%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|56|60|93.33%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AbstractOperation.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/MigrationContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/OperationDispatcher.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV1.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyTask.java|2|2|100.00%|

**Comment by amunra** [comment]
If I remove write permissions from the partition directory "force drop" states success, but leaves the directory intact.

**Comment by bluestreak01** [comment]
@amunra "alter" is likely written to WAL and picked up for execution asynchronously. This is my hunch. I would also check if it suspends the table or not if it fails.

**Comment by amunra** [comment]
> @amunra "alter" is likely written to WAL and picked up for execution asynchronously. This is my hunch. I would also check if it suspends the table or not if it fails. @ideoma could you comment?

Not quite..

It tries it sync first, but if it fails it just logs and reports success.

According to Alex, this is by design.

```java
                    engine.getPartitionOverwriteControl().notifyPartitionMutates(tableToken, timestamp, txn, 0);
                    if (!ff.unlinkOrRemove(other, LOG)) {
                        LOG.info()
                                .$("could not purge partition version, async purge will be scheduled [path=").$substr(pathRootSize, other)
                                .$(", errno=").$(ff.errno()).I$();
                        scheduleAsyncPurge = true;
                    }
```

**Review by amunra — APPROVED** [review]

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 135 / 142 (95.07%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/MetadataServiceStub.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV2.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|56|60|93.33%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|23|24|95.83%|
|:large_blue_circle:|io/questdb/cairo/TxWriter.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AbstractOperation.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperationBuilder.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/mig/MigrationContext.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/OperationDispatcher.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/UpdateOperatorImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ops/AlterOperation.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV1.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CopyTask.java|2|2|100.00%|

**Pull request merged**